packages/daemons/blocks_collection.go

```diff
@@ -184,64 +186,161 @@ func UpdateChain(ctx context.Context, d *daemon, host string, maxBlockID int64)
-		if err != nil {
-			// we got bad block and should ban this host
-			banNode(host, err)
-			d.logger.WithFields(log.Fields{"error": err, "type": consts.BlockError}).Error("processing block")
-			return err
-		}
+		rawBlocksCh <- rbr{blockID: blockID, rawBlock: blockBin}
+		semaphore <- true
+	}
 
-		// hash compare could be failed in the case of fork
-		hashMatched, thisErrIsOk := block.CheckHash()
-		if thisErrIsOk != nil {
-			d.logger.WithFields(log.Fields{"error": err, "type": consts.BlockError}).Error("checking block hash")
+	// rawBlocksQueue is 2nd conveyor stage. It storing blocks and send them in a sequential order to 2nd stage
+	rawBlocksQueue := func(rawBlocksCh chan rbr, rawBlocksQueueCh chan []byte, fbID int64) {
+		type rbb struct {
+			rbrs []rbr
+			*sync.RWMutex
 		}
 
-		if !hashMatched {
-			// it should be fork, replace our previous blocks to ones from the host
-			err := parser.GetBlocks(blockID-1, host)
+		nextBlockID := fbID
+		buffer := rbb{RWMutex: &sync.RWMutex{}}
+
+		go func() {
+			for rb := range rawBlocksCh {
+				buffer.Lock()
+				buffer.rbrs = append(buffer.rbrs, rb)
+				buffer.Unlock()
+			}
+		}()
+
+		go func() {
+			for {
+				buffer.RLock()
+				for _, orb := range buffer.rbrs {
+					if nextBlockID == orb.blockID {
+						rawBlocksQueueCh <- orb.rawBlock
+						nextBlockID = orb.blockID + 1
+					}
+				}
+				buffer.RUnlock()
+			}
+		}()
+	}
+
+	// playRawBlock is 3nd conveyor stage. It parses blocks and writing them into blockchain
+	playRawBlock := func(rawBlocksQueueCh chan []byte, errCh chan error, s *sync.WaitGroup) {
+		for rb := range rawBlocksQueueCh {
+			nb := make([]byte, len(rb))
+			copy(nb, rb)
+
+			block, err := parser.ProcessBlockWherePrevFromBlockchainTable(nb)
 			if err != nil {
-				d.logger.WithFields(log.Fields{"error": err, "type": consts.ParserError}).Error("processing block")
+				// we got bad block and should ban this host
 				banNode(host, err)
-				return err
+				d.logger.WithFields(log.Fields{"error": err, "type": consts.BlockError}).Error("processing block")
+				errCh <- err
+				return
 			}
-		} else {
-			/* TODO should we uncomment this ?????????????
-			_, err := model.MarkTransactionsUnverified()
+
+			// hash compare could be failed in the case of fork
+			hashMatched, thisErrIsOk := block.CheckHash()
+			if thisErrIsOk != nil {
+				d.logger.WithFields(log.Fields{"error": err, "type": consts.BlockError}).Error("checking block hash")
+			}
+
+			if !hashMatched {
+				// it should be fork, replace our previous blocks to ones from the host
+				err := parser.GetBlocks(block.Header.BlockID-1, host)
+				if err != nil {
+					d.logger.WithFields(log.Fields{"error": err, "type": consts.ParserError}).Error("processing block")
+					banNode(host, err)
+					errCh <- err
+					return
+				}
+			} else {
+				/* TODO should we uncomment this ?????????????
+				_, err := model.MarkTransactionsUnverified()
```

Comment:
no comments in sources please. If something is unclean - sak command and solve the problem